File: kotlin/grade-school/src/main/kotlin/School.kt
class School {
    private var db = sortedMapOf<Int, List<String>>()

    fun db(): Map<Int, List<String>> = db

    fun add(student: String, grade: Int) {
        db[grade] = grade(grade) + student
    }

    fun grade(grade: Int): List<String> = db.getOrElse(grade) { listOf() }

    fun sort(): Map<Int, List<String>> = db.mapValues { it.value.sorted() }
}

File: haskell/grade-school/School.hs
module School where

import Data.Function (on)
import Data.Ord (comparing)
import Data.List (sortBy, nubBy, sort)
import Data.Maybe (fromMaybe)

type School = [(Int, [String])]

empty :: School
empty = []

add :: Int -> String -> School -> School
add g name school = nubBy ((==) `on` fst) updatedSchool where
    updatedSchool = (g, updatedNames) : school
    updatedNames = sort $ name : grade g school

sorted :: School -> School
sorted school = sortBy (comparing fst) school

grade :: Int -> School -> [String]
grade g school = fromMaybe [] $ lookup g school